rfcs/0000-buildpack-resource.md

```diff
@@ -0,0 +1,218 @@
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Buildpack
+metadata:
+  name: my-custom-app-buildpack-0.3.1
+  namespace: my-kpack-using-namespace
+spec:
+  image: gcr.io/my-org/my-custom-app-buildpack:0.3.1
+  serviceAccountName: my-service-account
+```
+
+The status of Buildpack resource should include the same metadata to the ClusterStore. This is to enable kpack to calculate the exact expected digest of the Builder image without needing to communicate with the registry. Additionally, this metadata may be used by external tooling such as the [kp cli](https://github.com/pivotal/kpack). The status must include the id, buildpack api, diffID, digest, size, homepage, supported stacks, and order if a provided buildpack is a [meta-buildpacks](https://buildpacks.io/docs/concepts/components/buildpack/#meta-buildpack) of each buildpack within the configured buildpackage image.
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Buildpack
+metadata:
+  name: my-custom-app-buildpack-0.3.1
+  namespace: my-kpack-using-namespace
+spec:
+  image: gcr.io/my-org/my-custom-app-buildpack:0.3.1
+status:
+  conditions:
+  - status: "True"
+    type: Ready
+  buildpacks:
+  - api: "0.6"
+    diffId: sha256:05aed3622d52837de1b1950069f6cca10aa3dbf7429e5c0497bae8cee2e6d901
+    digest: sha256:9645b297c4c40b8476a243ae3c1ac9868986f64a000f9b95670ca9659428cec2
+    homepage: https://github.com/my-org/my-custom-app-buildpack
+    id: my-org/my-custom-app-buildpack
+    order:
+    - group:
+      - id: my-org/some-custom-implementation-buildpack
+        version: 1.4.0
+    size: 3342595
+    stacks:
+    - id: io.buildpacks.stacks.bionic
+    version: 4.2.2
+  - api: "0.6"
+    diffId: sha256:8e5c411f033cc771555aacbcb605cdbae7e3ae051a89515c4c278d0f704bd4ce
+    digest: sha256:62812731bf5264f78b93c024adcfa5942929bdc07357041f70ca5d360d96bc5c
+    id: my-org/some-custom-implementation-buildpack
+    size: 3373979
+    stacks:
+    - id: io.buildpacks.stacks.bionic
+    version: 1.4.0
+```
+
+
+##### Cluster Buildpack Resource
+
+```yaml 
+apiVersion: kpack.io/v1alpha3
+kind: ClusterBuildpack
+metadata:
+  name: paketo-java-buildpack-5.9.1
+spec:
+  image: gcr.io/paketo-buildpacks/java:5.9.1
+``` 
+
+A ClusterBuildpack should be available to be used within any ClusterBuilder and Builders in any namespace. Optional credentials  should provided to the ClusterBuildpack through a service account reference
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: ClusterBuildpack
+metadata:
+  name: paketo-java-buildpack-5.9.1
+spec:
+  image: privateregistry.com/paketo-buildpacks/java:5.9.1
+  serviceAccountRef:
+    name: private-sa
+    namespace: private-ns
+```
+
+The status of the ClusterBuildpack should match the status of the Buildpack resource.
+
+##### ClusterStore Resources
```

Comment:
Will there be any migration concerns?  One problem that jumped to mind is that this may increase complexity for getting started compared to configuring a ClusterStore.  Although the user has to configure one less resource, they can pretty much copy and paste several lines of paketo buildpacks in one yaml file to configure the clusterstore.  From this RFC it seems like the user would have to configure a buildpack/clusterbuildpack for each buildpackage they want to reference in a builder/clusterbuilder.

Reply:
> One problem that jumped to mind is that this may increase complexity for getting started compared to configuring a ClusterStore

I can add this to the RFC but, I hope that this will actually help simplify onboarding. Currently, because the Clusterstore is a monolithic resource users essentially need to use the  `kp` cli and a streamlined bootrap experience would require a published descriptor to be used alongside the cli. However, a Buildpack resource enables the Buildpacks to be managed independently which means a descriptor is not needed. Instead users can directly apply a preconfigured set of Buildpack/Builder/Stack resources with `kubectl apply` to get started fast. (This does require this WIP RFC to be implemented as well https://github.com/pivotal/kpack/pull/932)